rs/ledger_suite/icrc1/ledger/src/lib.rs

```diff
@@ -503,36 +545,80 @@ impl<Tokens: TokensType> Ledger<Tokens> {
 
         ledger
     }
+
+    pub fn is_ready(&self) -> bool {
+        matches!(self.state, LedgerState::Ready)
+    }
+
+    pub fn migrate_one_allowance(&mut self) -> bool {
+        match self.approvals.allowances_data.pop_first_allowance() {
+            Some((account_spender, allowance)) => {
+                self.stable_approvals
+                    .allowances_data
+                    .set_allowance(account_spender, allowance);
+                true
+            }
+            None => false,
+        }
+    }
+
+    pub fn migrate_one_expiration(&mut self) -> bool {
+        match self.approvals.allowances_data.pop_first_expiry() {
+            Some((timestamp, account_spender)) => {
+                self.stable_approvals
+                    .allowances_data
+                    .insert_expiry(timestamp, account_spender);
+                true
+            }
+            None => false,
+        }
+    }
+
+    pub fn clear_arrivals(&mut self) {
+        self.approvals.allowances_data.clear_arrivals();
+    }
 }
 
-impl<Tokens: TokensType> LedgerContext for Ledger<Tokens> {
+impl LedgerContext for Ledger {
     type AccountId = Account;
-    type AllowancesData = HeapAllowancesData<Self::AccountId, Tokens>;
+    type AllowancesData = StableAllowancesData;
     type BalancesStore = BTreeMap<Self::AccountId, Tokens>;
     type Tokens = Tokens;
 
     fn balances(&self) -> &Balances<Self::BalancesStore> {
+        if !self.is_ready() {
```

Comment:
Could we harden the check by including it also at every public entry point, i.e., update/query/composite query method, heartbeat, inspect message?

Reply:
Since `balances`, `approvals` and `stable_approvals` are private to the ledger and only accessed through getters or migration functions, no read or modification should happen during the migration. Other endpoints will return correct results even during the migration.

Reply:
Is there a chance that some of these traps would occur after an await in a public entry point (this is the motivation why blocking all public entry points might harden the check)?

Reply:
The only async endpoints are `icrc1_transfer`, `icrc2_transfer_from` and `icrc2_approve`. I added additional checks in those functions.

Reply:
> The only async endpoints

and the ledger code is not using stuff like `ic_cdk::spawn` to make async call from blocking code?

Reply:
As far as I know, this is only done when creating an archive after a transfer/approve. However, there are only 4 update endpoints in the ledger: the 3 above and `icrc21_canister_call_consent_message` which doesn't modify the ledger state and is only an update to make sure users get a reliable result (we could also disable it just in case). If we don't modify the state, could it be a problem if a `spawn` fails later, after the user got the result?

Reply:
> If we don't modify the state, could it be a problem if a spawn fails later, after the user got the result?

I didn't get the scenario, but ideally we don't invoke spawn at all during migration.

Reply:
I agree! In the ledger spawn is only called in the context of creating an archive, so it should not be called if no transfers or approvals are made.